File: Wire-iOS/Sources/Analytics/AnalyticsVoiceChannelTracker.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "AnalyticsVoiceChannelTracker.h"
#import "Analytics.h"
#import "WireSyncEngine+iOS.h"
#import "DeveloperMenuState.h"


@interface AnalyticsVoiceChannelTracker () <VoiceChannelStateObserver>

@property (nonatomic) BOOL initiatedCall;
@property (nonatomic) BOOL isVideoCall;
@property (nonatomic) NSDate *callEstablishedDate;
@property (nonatomic) NSDate *callAnsweredDate;

@property (nonatomic) Analytics *analytics;
@property (nonatomic) id voiceChannelStateObserverToken;

@end



@implementation AnalyticsVoiceChannelTracker

- (instancetype)initWithAnalytics:(Analytics *)analytics
{
    self = [super init];
    
    if (self) {
        self.analytics = analytics;
        self.voiceChannelStateObserverToken = [VoiceChannelV3 addStateObserver:self userSession:[ZMUserSession sharedSession]];
    }
    
    return self;
}


#pragma mark - VoiceChannelStateObserver

- (void)callCenterDidChangeVoiceChannelState:(VoiceChannelV2State)voiceChannelState conversation:(ZMConversation *)conversation callingProtocol:(enum CallingProtocol)callingProtocol
{
    if (voiceChannelState == VoiceChannelV2StateOutgoingCall) {
        self.initiatedCall = YES;
        self.isVideoCall = conversation.voiceChannel.isVideoCall;
        [self.analytics tagInitiatedCallInConversation:conversation video:self.isVideoCall callingProtocol:callingProtocol];
    }
    else if (voiceChannelState == VoiceChannelV2StateIncomingCall) {
        self.initiatedCall = NO;
        self.isVideoCall = conversation.voiceChannel.isVideoCall;
        [self.analytics tagReceivedCallInConversation:conversation video:self.isVideoCall callingProtocol:callingProtocol];
    }
    else if (voiceChannelState == VoiceChannelV2StateSelfIsJoiningActiveChannel) {
        self.callAnsweredDate = [NSDate date];
        [self.analytics tagJoinedCallInConversation:conversation video:self.isVideoCall initiatedCall:self.initiatedCall callingProtocol:callingProtocol];
    }
    else if (voiceChannelState == VoiceChannelV2StateSelfConnectedToActiveChannel && nil == self.callEstablishedDate) {
        self.callEstablishedDate = [NSDate date];
        NSTimeInterval setupDuration = [self.callEstablishedDate timeIntervalSinceDate:self.callAnsweredDate];
        [self.analytics tagEstablishedCallInConversation:conversation video:self.isVideoCall initiatedCall:self.initiatedCall setupDuration:setupDuration callingProtocol:callingProtocol];
    }
}

- (void)callCenterDidFailToJoinVoiceChannelWithError:(NSError *)error conversation:(ZMConversation *)conversation
{
    
}

- (void)callCenterDidEndCallWithReason:(VoiceChannelV2CallEndReason)reason conversation:(ZMConversation *)conversation callingProtocol:(enum CallingProtocol)callingProtocol
{
    if (reason == VoiceChannelV2CallEndReasonInputOutputError && [DeveloperMenuState developerMenuEnabled]) {
        UIAlertView* view = [[UIAlertView alloc] initWithTitle:@"Calling error"
                                                       message:@"AVS I/O error"
                                                      delegate:nil
                                             cancelButtonTitle:@"OK"
                                             otherButtonTitles:nil];
        [view show];
    }
    [self.analytics tagEndedCallInConversation:conversation
                                         video:self.isVideoCall
                                 initiatedCall:self.initiatedCall
                                      duration:-[self.callEstablishedDate timeIntervalSinceNow]
                                        reason:reason
                               callingProtocol:callingProtocol];
    self.callAnsweredDate = nil;
    self.callEstablishedDate = nil;
}

@end

File: Wire-iOS/Sources/Analytics/AnalyticsVoiceChannelTracker.swift
//
// Wire
// Copyright (C) 2017 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import Foundation

fileprivate let zmLog = ZMSLog(tag: "calling")

struct CallInfo {
    var answeredDate : Date?
    var establishedDate : Date?
    let outgoing : Bool
}

class AnalyticsVoiceChannelTracker : NSObject {
    
    let analytics : Analytics
    var callInfos : [UUID : CallInfo] = [:]
    var callStateObserverToken : Any?
    
    init(analytics : Analytics) {
        self.analytics = analytics
        
        super.init()
        
        guard let userSession = ZMUserSession.shared() else {
            zmLog.error("UserSession not available when initializing \(type(of: self))")
            return
        }
        
        self.callStateObserverToken = WireCallCenterV3.addCallStateObserver(observer: self, userSession: userSession)
    }
}

extension AnalyticsVoiceChannelTracker : WireCallCenterCallStateObserver {
    
    func callCenterDidChange(callState: CallState, conversation: ZMConversation, user: ZMUser?, timeStamp: Date?) {
        
        let conversationId = conversation.remoteIdentifier!
        
        switch callState {
        case .outgoing:
            analytics.tagInitiatedCall(in: conversation, video: conversation.voiceChannel?.isVideoCall ?? false)
            callInfos[conversationId] = CallInfo(answeredDate: nil, establishedDate: nil, outgoing: true)
        case .incoming(video: let video, shouldRing: _, degraded: _):
            analytics.tagReceivedCall(in: conversation, video: video)
            callInfos[conversationId] = CallInfo(answeredDate: nil, establishedDate: nil, outgoing: false)
        case .answered:
            if let callInfo = callInfos[conversationId] {
                analytics.tagJoinedCall(in: conversation, video: conversation.voiceChannel?.isVideoCall ?? false, initiatedCall: callInfo.outgoing)
                callInfos[conversationId] = CallInfo(answeredDate: Date(), establishedDate: nil, outgoing: callInfo.outgoing)
            }
        case .established:
            if let callInfo = callInfos[conversationId] {
                let video = conversation.voiceChannel?.isVideoCall ?? false
                let setupDuration = -(callInfo.answeredDate?.timeIntervalSinceNow ?? -999)
                analytics.tagEstablishedCall(in: conversation, video: video, initiatedCall: callInfo.outgoing, setupDuration: setupDuration)
                callInfos[conversationId] = CallInfo(answeredDate: callInfo.answeredDate, establishedDate: Date(), outgoing: callInfo.outgoing)
            }
            
        case .terminating(reason: let reason):
            if let callInfo = callInfos[conversationId] {
                let video = conversation.voiceChannel?.isVideoCall ?? false
                let duration = -(callInfo.establishedDate?.timeIntervalSinceNow ?? -0)
                analytics.tagEndedCall(in: conversation, video: video, initiatedCall: callInfo.outgoing, duration: duration, callEndReason: reason.analyticsValue)
            }
            callInfos[conversationId] = nil
            
            if case .inputOutputError = reason {
                presentIOErrorAlertIfAllowed()
            }
        default:
            break
        }
        
    }
    
    func presentIOErrorAlertIfAllowed() {
        guard DeveloperMenuState.developerMenuEnabled() else { return }
        
        let alert = UIAlertController(title: "Calling Error", message: "AVS I/O error", cancelButtonTitle: "OK")
        alert.presentTopmost()

    }
    
}

private extension CallClosedReason {
    
    var analyticsValue : String {
        switch self {
        case .canceled:
            return "canceled"
        case .normal, .stillOngoing:
            return "other" // Current API doesn't expose if we or the remote ended the call
        case .inputOutputError:
            return "io_error"
        case .internalError:
            return "internal_error"
        case .anweredElsewhere:
            return "answered_elsewhere"
        case .timeout:
            return "timeout"
        case .unknown:
            return "unknown"
        case .lostMedia:
            return networkQualityString()
        }
    }
    
    func networkQualityString() -> String {
        let qualityType = NetworkConditionHelper.sharedInstance().qualityType()
        
        switch qualityType {
        case .typeWifi:
            return  "drop_wifi"
        case .type2G:
            return "drop_2G"
        case .type3G:
            return "drop_3G"
        case .type4G:
            return "drop_4G"
        case .typeUnkown:
            return "drop_unknown"
        }
    }
    
}
